beps/0002: add dynamic FE plugins cli,backend,loading proposals (#24915)
backstage/backstage: beps/0002-dynamic-frontend-plugins/README.md

A simple program to traverse the AST and find exports from the `main` or `module` script of a dependency. Any exposed modules from the dependency entry will be traced to their source files, and a module map can be generated.

> NOTE: It is important that import paths in source code are also directly targeting the "traced" modules or are transformed at build time.

```sh
backstage-cli dependency generate-module-path --root <path-to-dependency-root>
```

Generates a module map (exported module source file location) for module sharing.

**Arguments**

- _path_
  - path to the module root (directory with package.json)

[module map generator using TS example](#simplistic-minimal-example-of-module-map-generator)

### Dynamic assets server plugin

A simple static asset server is sufficient to serve the dynamic assets. The asset server can collect the necessary metadata at boot or runtime.

#### Plugin separation

- Each plugin should be separated into a directory.

> - Each plugin will require its dependencies in the `package.json` to be installed if its supposed to be "dynamized".

Author: Agreed. I'd also expect installed plugins to be dynamic already. Should we say this is out of scope and handled as a part of the installation process? Make a hard requirement for the plugin to be already dynamic.

> - Each plugin will require its dependencies in the `package.json` to be installed if its supposed to be "dynamized".

Reviewer: Yeah I wouldn't mind making that a requirement for now and adding this as an optional feature down the line.

> - Installed node modules **can't be shared** to avoid dependency conflicts. Module federation is able to handle duplicate dependency versions.

Reviewer: Aight, let's keep as is, gonna resolve this

- Each plugin can be viewed as a `private` workspace package in a npm workspaces environment.

```sh
path/to/assets/plugin/dynamic-plugins/
├── dynamic-backstage.plugin-foo
│   ├── dist
│   ├── node_modules # private_modules for plugins that need to be dynamized!
│   ├── package.json
│   ├── package-lock.json
│   └── plugin-manifest.json
└── dynamic-backstage.plugin-user-settings
    ├── dist
    ├── package.json
    └── plugin-manifest.json
```

#### Boot time initialization

**Dynamically build plugins**

Plugins that already contain dynamic output. Including the [manifest](#plugin-manifest). No additional init required.

**Static plugins**

Have to be ["Dynamized"](#dynamize-plugin-during-backstage-boot) first. Plugin dependencies has to be installed. The new [Dynamic export CLI](#dynamic-export-cli) has to be called first.

> **Plugin registry collection**

Author: > or may be it's even just an ingestion method and we have something else for the runtime storage

Probably this. Doe backstage already have some storage alternatives?

> **Plugin registry collection**

Reviewer: Yep, there's DB and cache, and it's somewhat likely that we'd add some form of object store in not too long either.

**Singleton dependencies verification (optional)**

With [module sharing](#module-sharing) and its singleton option, it is critical that the singleton dependencies versions are matching.

The singleton configuration from the shell application can be extracted as well as from all dynamic plugins (included in the generated [manifest](#plugin-manifest)). A semver check can be used to prevent a plugin from initializing if a version mismatch is found. A version check also has to be made between plugins. New plugin singleton dependencies must match both the shell and all previously loaded plugins.

#### Runtime updates

It is fully expected that plugins can be added or removed at the asset server runtime. The same tasks can be performed as during boot time initialization. Assigning a FS watcher (like [Chokidar](https://github.com/paulmillr/chokidar)) can be used to handle updates.

#### Asset server API

> NOTE Routes names are purely arbitrary at this point.

The API can be very simplistic

_GET /dynamic-plugins/plugin-registry_

Returns generated [Plugin Registry](#plugin-registry-1).

> \*GET /dynamic-plugins/assets/\*\*

Author: We had similar discussions about bundles. I think "bundles" should be just an abstraction layer. It would be up to the plugin builders to define "bundles" or dependencies.

The question is, what is supposed to happen with a plugin if a dependency is missing? All plugins are loaded at boot time (in the frontend) technically all dependencies should be already present in the backend. I think the concept of bundles is just a cosmetic grouping for some plugin marketplace rather than something necessary for the API itself. At most, the API should only validate if a plugin dependency is also available.

          // the init scripts URL is enhanced with the asset location to add the host:
          return `${manifest.customProperties.assetHost}${script}`;
        }),
      };
    },
  },
};

// Initialize Scalprum instance
const scalprum = initialize({
  appsConfig: pluginRegistry,
  pluginLoaderOptions: pluginSDKOptions.pluginLoaderOptions,
});
```

Because the [`appLoader`](https://github.com/backstage/backstage/blob/master/packages/frontend-app-api/src/wiring/createApp.tsx#L193) is already async, it is a perfect place to load the plugin registry and init the dynamic plugins.

Initializing the dynamic feature is just a case of mapping the `DynamicFrontendFeature` to `FrontendFeature` via Scalprum:

```ts
import { processManifest, getModule } from '@scalprum/core';
// a ID of the module withing module federation container, can be customized, depends on the build
const DEFAULT_MODULE_NAME = 'pluginEntry';

async function loadScalprumFeature({ manifestLocation, name }) {
  // initialize dynamic remote within scalprum cache
  await processManifest(manifestLocation, name, DEFAULT_MODULE_NAME);
  // gets the actual JS module from scalprum

>   const feature = await getModule<FrontendFeature>(name, DEFAULT_MODULE_NAME);

Author: Can be used as well. The proposed implementation already can handle manifest files, has caching building, can resolve non-default import names etc.

  return feature;
}
const providedFeatures: FrontendFeature[] = [];
for (const entry of options?.features ?? []) {
  if ('manifestLocation' in entry && 'name' in entry) {
    // some error handling required
    const dynamicFeature = loadScalprumFeature(entry);
    providedFeatures.push(dynamicFeature);
  } else if ('load' in entry) {
    // plugin with load function
  } else {
    // #static plugin
  }
}

const app = createSpecializedApp({
  icons: options?.icons,
  config,
  features: [...discoveredFeatures, ...providedFeatures],
  bindRoutes: options?.bindRoutes,
}).createRoot();
```

The federated modules are now available as if they were static plugins.

> **Access to scalprum provider**

Author: I think benefits would come when you start truly lazy loading plugins. Meaning not having to initialize them at boot time. You'd also have the ability to access the extensions from the SDK (the part that is closest to current backstage plugins).

With the current plan, no benefits are having the react binding accessible as you are not using them.

> **Access to scalprum provider**

Reviewer: ah alright, could we leave this out of scope for now and remove it from the proposal?

#### Static and Dynamic Plugin Build Differences

The major difference between static and dynamic plugins lies in their dependencies and how they are "installed" into a Backstage instance and whether they are bundled into final JS output during the build.

While static plugins can define their dependencies using the dependencies field inside `package.json`, dynamic (module federation-based) plugins can't.

A dynamic plugin requires all dependencies to be present in the build output. Module sharing is then used to optimize the "output size" at runtime. If a shared package is not found in the browser scope, or if the version range does not match the requirement, it falls back and loads additional JS assets from the "private" build and adds them to the scope.

Webpack is also not the best tool for building traditional NPM packages. Dependencies can be externalized, but Webpack brings additional code into the bundles. Using Webpack external dependencies also conflicts with module federation shared modules. Externals will not be included even as a shared module fallback. As a result, such shared module can't be used because the dependency will not be defined in the init container.

> If the vision is to still support "static" plugins installed via NPM, it makes sense to keep the static and dynamic plugin builds separate and build two different bundles. Both can still be published as a single NPM package (not recommended), but to achieve optimal performance for both, there cannot be a shared output.

Reviewer: Got a bit nerd-sniped here thinking: what if we could do exactly this? Having a single build that is compatible with both dynamic and build-time bundling? When bundling the app you'd simply ignore all of the bundled transitive dependencies. As you mention I think it would require some very ugly solution, but it does have me intrigued 🤔 😁

Author: I was trying to think of a solution that would allow us exactly this. The problem is the dependencies though :face_exhaling: .

> what if we could do exactly this?

I think it can be done. But it is not trivial. We might have to dig into some custom webpack plugins or even worse. I don't think its worth the effort **for now**. I'd consider this as a follow up.

The fact that the static plugins are supposed to be your traditional NPM packages is the problem here. Webpack is a must have for module federation. We have to bundle dependencies. Which is exactly what you don't want for NPM dependencies. Externals can't be used because it messes up the shared scopes. This problem does not exist in a world where the plugins do not depend on NPM ecosystem.

I think (maybe it's just a hope) that eventually, dynamic plugins will be the default way of using backstage plugins. And instead of NPM there will be an ecosystem of CNDs (like RHDH, roadiehq, Spotify) that will distribute their dynamic plugins.

#### Dynamic export CLI

```sh

> backstage-cli plugin export-dynamic

Reviewer: Another option to consider is to make it a flag instead, something like `build --dynamic`. Mainly thinking that could be useful because it might work well with our existing `repo build` commands etc. Fewer additions overall. Not super important for the overall solution at this point though.

Author: Agreed I had these two options in mind.

```

Similar to the current build command, instead of producing static plugin output, it builds a dynamic plugin using Webpack and module federation plugins.

**Arguments**

- _path_ (optional)
  - path to the plugin `package.json`
  - defaults to `cwd`
- _export-name_
  - Export name from where the plugin root is accessible
  - defaults to `pluginEntry`
- _cwd_
  - current working directory of the script

>   - defaults to current `cwd` value

Reviewer: I think we can have all of these hardcoded tbh. The script always targets `cwd`, that's how the the existing commands work. The path is always `package.json` or defined inside it. The export is always the default one, which is already part of the new frontend and backend systems.

**Shared modules**

> NOTE: Details of module sharing are described in the [Module sharing](#module-sharing) section.

Short summary of dependency sharing:

- Predefined set of singleton packages
- Generated granular config of modules for MUI dependencies
- Share all other dependencies listed in package.json

> To allow additional customization of module sharing, extra configuration will be required. A similar approach to the current catalog-info.yaml config file can be used. There is also the option of embedding the metadata in the `package.json`. However, the configuration can become very large and clutter the package.json.

Reviewer: Do you have an example in mind where a significant amount of config would be required?

As a side note, the existing `catalog-info.yaml` files don't really have anything to do with the Backstage build tooling, it's used for tracking in the catalog and shouldn't be used as an excuse for adding more YAML 😁

Author: > the existing catalog-info.yaml files don't really have anything to do with the Backstage build tooling,

I was not aware of that. 

> Do you have an example in mind where a significant amount of config would be required?

I think whenever some component library (like MUI) is present in a plugin. I also think at some point plugin builders might want to exclude specific dependencies from sharing. I personally don't like adding anything to `package.json` that is not related to npm (I've seen to many jest configs taking over the whole file). But that's my personal preference. If you prefer using the file, I don't have any objective argument why not.

Author: I'd say typically you would not change the defaults at all. The vast majority should be covered by the default settings. I'll change the section and propose some package.json entries. Do you already have some backstage build-related fields reserved in `package.json`?

- [Schema proposal draft](./dependency-sharing-schema.json)
- [Config draft](./share-config-sample.yaml)

#### Generate module sharing map

> > NOTE: RedHat has similar internal generators that are used in combination with Scalprum. These can be built upon, moved to Scalprum, and exposed. If there is interest, Spotify does not have to be responsible for maintaining the source code.

Reviewer: ```suggestion
> NOTE: RedHat has similar internal generators that are used in combination with Scalprum. These can be built upon, moved to Scalprum, and exposed. If there is interest, the Backstage maintainers do not have to be responsible for maintaining the source code.
```

> A simple program to traverse the AST and find exports from the `main` or `module` script of a dependency. Any exposed modules from the dependency entry will be traced to their source files, and a module map can be generated.

Reviewer: Is it important enough to find modules that are truly imported as opposed to simply including all transitive `package.json` dependencies for this?

Author: Most package.json dependencies do not matter and can be shared as a whole. There is no point looking for exports for `react` or `react-router-dom`. In fact, we do not share as a whole package, the module sharing breaks.

However specifically for MUI packages (and any other component libraries), it is critical to target specifically modules that are used. Sharing the whole shebang means the entire thing ends up in the bundle (can't tree share shared module). MUI icons package has `21226` modules (at the time of writing). That's 9.1 megabytes of JS. You have another 8 for the `@mui/material`. AFAIK, you also have the older version of MUI dependencies. When you load 4-10 (educated guess, way more for pinned dependencies) versions in a session where the sharing is not optimized you can have hundreds of megabytes of JS just for some icons and buttons. In that scenario, it is better not to share the dependency.

Author: Yes exactly. The AST I am proposing generates the module map. I see no harm in exposing the generator so it can be used for other dependencies as well. I don't think we want developers to construct MUI like config manually.

> NOTE: It is important that import paths in source code are also directly targeting the "traced" modules or are transformed at build time.

```sh
backstage-cli dependency generate-module-path --root <path-to-dependency-root>
```

> Generates a module map (exported module source file location) for module sharing.

Reviewer: This strikes me as something that should be built into the above export command?

Author: It should triggered by it, and might not have to be exposed as a CLI script, but the code should definitely be separate.

**Arguments**

- _path_
  - path to the module root (directory with package.json)

> **Simplistic minimal example of module map generator**

Reviewer: Let's move this down to the design details section 😁

  }
  const typeChecker = program.getTypeChecker();

  /** @param {typescript.SourceFile} sourceFile */
  const getExportNames = (sourceFile) => typeChecker.getExportsOfModule(typeChecker.getSymbolAtLocation(sourceFile)).map((symbol) => symbol.getName());

  const baseExports = getExportNames(program.getSourceFile(base));
  const compilerDir = compiler.getCurrentDirectory();
  const resolvedModules = baseExports.map((name) => {
    const res = typescript.resolveModuleName(name, base, defaultCompilerOptions, compiler, moduleResolutionCache);
    if(res.resolvedModule?.resolvedFileName) {
      return ({
        name,
        file: libName + res.resolvedModule.resolvedFileName.replace(compilerDir, '').replace(/\.js$/, '')
      })
    }
    return undefined
  }).filter((x) => x !== undefined);

  const moduleMapFileName = path.resolve(__dirname, 'module-map.json');
  fs.writeFileSync(moduleMapFileName, JSON.stringify(resolvedModules, null, 2));
}

const libName = "@mui/foo";

getDynamicModuleMap(libName);

/***
 * [
 *  {
 *    "name": "foo",
 *    "file": "@mui/foo/dist/foo"
 *  },
 *  {
 *    "name": "baz",
 *    "file": "@mui/foo/dist/baz"
 *  }
 * ]

>  * /

Reviewer: ```suggestion
 */
```

  let pluginName: string;
  if (plugin.name.includes('/')) {
    const fragments = plugin.name.split('/');
    pluginName = `${fragments[0].replace('@', '')}.${fragments[1]}`;
  } else {
    pluginName = plugin.name;
  }

  return `dynamic-${pluginName}`;
}

// plugin = { name: "@backstage/plugin-api-docs", version: "1.0.0" } based on package.json
const pluginName = getName(plugin);

const { ModuleFederationPlugin, ContainerPlugin } = require('@module-federation/enhanced');

const dynamicPluginPlugin = new DynamicRemotePlugin({
  extensions: [],
  sharedModules, // from shared config and generated from `package.json` dependencies
  entryScriptFilename: `[name].[contenthash].js`, // using contenthash for cashing purposes
  moduleFederationSettings: {
    libraryType: 'global', // use the globalThis object rather than jsonp
    // instruct the sdk plugin to use @module-federation/enhanced tooling
    pluginOverride: {
      ModuleFederationPlugin,
      ContainerPlugin,
    }
  },
  pluginMetadata: {
    name: pluginName,
    version: plugin.version || '0.0.0',
    exposedModules: {
      // path to the default export of the frontend plugin entry point

>       pluginEntry: './src/index.js',

Author: It should point to source files, not transpiled bundles. `main` or `module` fields are usually not pointing to source files. If we are talking about rebuilding the existing package, you should create a fake entry point that re-reexports the plugin entry (default).

>       pluginEntry: './src/index.js',

Reviewer: `main` points to a source file with a Backstage build setup, e.g. https://github.com/backstage/backstage/blob/b6f2d3464717a0adc4936cd4582a4c63f9f3a13e/plugins/catalog/package.json#L27

>       pluginEntry: './src/index.js',

Reviewer: Yep as long as it's using `@backstage/cli`, which is the scope of this solution anyway

    },
  },
});

const config: Configuration = {
    context: pluginRoot,
    output: {
      chunkFilename: `${pluginName}.[contenthash].js`,
      path: path.resolve(pluginRoot, 'dist'),
      publicPath: 'auto',
    },
    entry: {},
    resolve: {
      extensions: ['.js', '.jsx', '.ts', '.tsx'],
    },
    plugins: [dynamicPluginPlugin],
    module: {
      rules: [
        {
          // needs TS config as well
          test: /\.js$/,
          exclude: /(node_modules)/,
          use: {
            loader: 'swc-loader',
          },
        },
        {
          test: /\.(png|svg|jpg|jpeg|gif)$/i,
          type: 'asset/resource',
        },
        // additional loaders for CSS if necessary, can use the same as the current backstage shell app
      ],
    },
  };
  return config;
```

### Plugin manifest

Plugin manifest is a simple JSON file containing critical metadata to initialize a remote plugin.

The `@openshift/dynamic-plugin-sdk-webpack` plugin will output manifest file in the following format:

```ts
type PluginManifest = {
  name: string;
  version: string;
  dependencies?: Record<string, string>;
  customProperties?: AnyObject;
  baseURL: string;

>   extensions: Extension[];

Author: It is the extension definition Vojtech was showing on the call we had. It can be empty and will be probably at least initially. Unless you decide it would be worth to adopt the concept.

>   extensions: Extension[];

Reviewer: Alright, could we drop it from the update for now?

>   extensions: Extension[];

Author: Sure thing :+1:

  loadScripts: string[];
  registrationMethod: 'callback' | 'custom';
  buildHash?: string;
};
```

This manifest is directly usable in Scalprum at runtime.

> **baseURL**

Author: The manifest URL does not matter. Manifests can be hard coded in the source code.

The base URL is determined from where the initial script is fetched. And that is not something you can guess when you are building the script. The script does not have to have the same origin as the manifest. I don't think backstage plugins will live on different origins initially, but that can change if some model of CDNs with plugins pops up.

I think there is no harm in proving a very explicit manifest.

> **baseURL**

Reviewer: It's best if the manifest/script is portable, as in you don't need to build it with any configuration that's specific to the environment that you want to host it at. This strikes me as something that could optionally be present for advanced deployment scenarios, but in a standard setup it's not needed?

> **baseURL**

Author: > This strikes me as something that could optionally be present for advanced deployment scenarios, but in a standard setup it's not needed?

Agree.
